### PR TITLE
[BEAM-2301] Splits SplittableParDo into a core-construction part and a runners-core part

### PR DESCRIPTION
**Pull request by jkff**
SplittableParDo itself goes into core-construction, and expands into a slightly different transform.

This change is almost completely simply moving code around.

Before:
```
elements: InputT
| pair with restriction -> ElementAndRestriction<InputT, RestrictionT>
| split restriction -> same
| explode windows -> same
| assign unique key -> KV<String, ElementAndRestriction<InputT, RestrictionT>>
| GBKIntoKeyedWorkItems -> KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>
| ProcessElements -> PCollection<OutputT>
```

After:
```
elements: InputT
| ...
| assign unique key -> KV<String, ElementAndRestriction<InputT, RestrictionT>>
| SplittableProcessKeyed -> PCollection<OutputT>
```

Most runners (except Dataflow) will still want to go through KeyedWorkItem. That part is encapsulated in `SplittableParDoViaKeyedWorkItems`, which has an `OverrideFactory` for `SplittableProcessKeyed` expanding it into the good old `GBKIntoKeyedWorkItems` and `ProcessElements`. So runner changes are very minor.

Dataflow, however, can not use runners-core during expansion, so it will translate `SplittableProcessKeyed` directly and perform its expansion service-side, and will instantiate `ProcessFn` worker-side.

R: @tgroh

### CODE REVIEW

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    }
+
+    @Override
+    public PCollectionTuple expand(
+        PCollection<KV<String, ElementAndRestriction<InputT, RestrictionT>>>
+            input) {
+      return createPrimitiveOutputFor(input);
+    }
+
+    public PCollectionTuple createPrimitiveOutputFor(PCollection<?> input) {
```
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java]: I'm not super happy with this being an instance variable. Even if you need to expose it (which I understand why it is currently), I'd rather make it a static method that takes in the `fn` as well.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            .apply(
+                "Explode windows",
+                ParDo.of(new ExplodeWindowsFn<ElementAndRestriction<InputT, RestrictionT>>()))
+            .apply(
+                "Assign unique key",
+                WithKeys.of(new RandomUniqueKeyFn<ElementAndRestriction<InputT, RestrictionT>>()));
+
+    return keyedRestrictions.apply(
+        "SplittableProcessKeyed",
+        new SplittableProcessKeyed<>(
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java]: That's pretty much what is going to happen in Dataflow runner. Other runners might later adopt a similar approach, but for now there seems no harm in letting them use runners-core for convenience.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    public void process(ProcessContext c, BoundedWindow window) {
+      c.output(c.element());
+    }
+  }
+
+  /**
+   * Runner-specific primitive {@link PTransform} that invokes the {@link DoFn.ProcessElement}
+   * method for a splittable {@link DoFn}.
+   */
+  public static class SplittableProcessKeyed<InputT, OutputT, RestrictionT>
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java]: Renamed to ProcessKeyedElements.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        DoFnSignatures.getSignature(parDo.getFn().getClass()).processElement().isSplittable(),
+        "fn must be a splittable DoFn");
+  }
+
+  @Override
+  public PCollectionTuple expand(PCollection<InputT> input) {
+    return applyTyped(input);
+  }
+
+  private PCollectionTuple applyTyped(PCollection<InputT> input) {
```
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java]: Can this be inlined? You don't introduce any type variables.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+ * A utility transform that executes a <a
+ * href="https://s.apache.org/splittable-do-fn">splittable</a> {@link DoFn} by expanding it into a
+ * network of simpler transforms:
+ *
+ * <ol>
+ * <li>Pair each element with an initial restriction
+ * <li>Split each restriction into sub-restrictions
+ * <li>Assign a unique key to each element/restriction pair
+ * <li>Group by key (so that work is partitioned by key and we can access state/timers)
+ * <li>Process each keyed element/restriction pair with the splittable {@link DoFn}'s {@link
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java]: Rephrased.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java (hunk 4 above)
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java]: Done.

**Pull request merged**